File: SIMD_exercise/Fortran/sin_of_vector.f90
Program sin_of_vector

  use trig_taylor
  !$ use omp_lib

  implicit none

  integer, parameter :: nsize = 8192
  double precision, dimension(nsize) :: x, y
  
  integer :: index
  double precision :: start_time, end_time

  integer :: repeat
  integer, parameter :: n_repeat = 1000
  
  ! intialise the vector
  Do index = 1, nsize
     x(index) = 0.001D0 * dble(index)
  end do

  !$ start_time = omp_get_wtime()
  Do repeat = 1, n_repeat
     ! calculate the sin
     !$omp simd 
     Do index = 1, nsize
        y(index) = sin_taylor(x(index)) 
     end do
  enddo

  !$ end_time = omp_get_wtime()  
  
  ! write the result
  open(11,file="sin_vector.out")
  do index = 1, nsize
     write(11,*) x(index), y(index)
  end do
  close(11);

  print *, "Calculation finished"
  !$ print *, "Calculation for n=", nsize, " took", &
  !$     (end_time - start_time)/n_repeat, " seconds"

End program sin_of_vector

File: SIMD_exercise/C/sin_of_vector.c
#include <stdio.h>
#include <stdlib.h>
#include "trig_taylor.h"
#include <omp.h>

int main()
{

  const int nsize = 8192;
  const int n_repeat = 1000;
  
  double* x = aligned_alloc(64, nsize* sizeof *x);
  double* y = aligned_alloc(64, nsize* sizeof *x);

  for (int i=0; i<nsize; i++)
    x[i] = 0.001 * (double)i;

#ifdef _OPENMP
  double start_time = omp_get_wtime();
#endif

  for (int t=0; t< n_repeat; t++)
  {
  // calculate the sin
  for (int i=0; i<nsize; i++)
    y[i] = sin_taylor(x[i]);
  }
#ifdef _OPENMP
  double end_time = omp_get_wtime();
#endif

  // write the result
  FILE * fp;
  fp = fopen("sin_result.out", "w");
  for(int i=0; i<nsize; i++)
    fprintf(fp, "%8.4f %12.9f\n", x[i], y[i]);
  fclose;

#ifdef _OPENMP
  printf("Calculation for n= %d took %f10 seconds\n", nsize, (end_time-start_time)/n_repeat);
#endif

  free(x);
  free(y);
  return 0;

}
